src/vs/workbench/api/common/extHostNotebookKernels.ts

```diff
@@ -404,6 +413,28 @@ export class ExtHostNotebookKernels implements ExtHostNotebookKernelsShape {
 		}
 	}
 
+	async $provideVariables(handle: number, requestId: string, notebookUri: UriComponents, parentName: string | undefined, kind: 'named' | 'indexed', start: number, token: CancellationToken): Promise<void> {
+		const obj = this._kernelData.get(handle);
+		if (!obj) {
+			return;
+		}
+
+		const document = this._extHostNotebook.getNotebookDocument(URI.revive(notebookUri));
+		const variableProvider = obj.controller.variableProvider;
+		if (!variableProvider) {
+			return;
+		}
+
+		const parent = parentName ? { name: parentName, value: '' } : undefined;
+		const variables = variableProvider.provideVariables(document.apiNotebook, parent, kind, start, token);
+		for await (const variable of variables) {
+			if (token.isCancellationRequested) {
+				return;
+			}
+			this._proxy.$receiveVariable(requestId, variable);
```

Comment:
If you're just going to pull all the variables here, then it seems like you could just return the variables from this call directly rather than send them over one by one?

Reply:
This could stream them into the renderer process asynchronously. 
Would there still a purpose to using an asyncIterable in the API if they just are gathered up at once in the EH anyway?

Reply:
I suppose it would make sense to send them all at once if I'm going to stick with the non-async tree type - I wasn't really settled on that yet.

Reply:
I figured that getting the next variable from the call wouldn't typically be slow, so if there are 10 variables they are going to basically appear all at once anyway, but that you could implement something like paging with this, eg you pull the first 10 variables out of here, display those, and pull more out later if needed. And also for the large array paging thing. This is fine for now of course. I think it confused me that this switches from a "pull model" (the iterable) to a push model (receiveVariable) and back.